core/controllers/load_functions.py

```diff
@@ -0,0 +1,233 @@
+    """
+    Loads data into the 'Outputs_Ref' or 'Outcomes_Ref' tables.
+
+    The function first retrieves the relevant data from the workbook using the provided mapping.
+    It then processes the data to get outcomes and outputs to insert,
+    and finally adds all the models to the database session.
+
+    :param workbook: a dictionary of DataFrames of table data to be inserted into the db.
+    :param mapping: the mapping of the relevant DataFrame to its attributes as they appear in the db.
+    """
+    worksheet = workbook[mapping.table]
+    models = mapping.map_data_to_models(worksheet)
+
+    models = get_outcomes_outputs_to_insert(mapping, models)
+    db.session.add_all(models)
+
+
+def generic_load(workbook: dict[str, pd.DataFrame], mapping: DataMapping, submission_id):
+    """
+    Generic function to load data into a specified table.
+
+    The function takes a workbook, a mapping of relevant data, and a submission ID.
+    It adds the submission ID to the worksheet, maps the data to models using the provided mapping,
+    and adds all the models to the database session.
+
+    :param workbook: a dictionary of DataFrames of table data to be inserted into the db.
+    :param mapping: the mapping of the relevant DataFrame to its attributes as they appear in the db.
+    :param submission_id: the ID of the submission associated with the data.
+    """
+    worksheet = workbook[mapping.table]
+    worksheet["Submission ID"] = submission_id
+    models = mapping.map_data_to_models(worksheet)
+
+    db.session.add_all(models)
+
+
+def delete_existing_submission(submission_to_del: str) -> None:
+    """
+    Deletes the existing submission and all its children based on the UUID of that submission.
+
+    :param submission_to_del: string of Submission's id to be deleted.
+    :return: None
+    """
+    Submission.query.filter_by(id=submission_to_del).delete()
+    db.session.flush()
+
+
+def get_or_generate_submission_id(workbook: dict[str, pd.DataFrame]) -> tuple[str, Submission | None]:
+    """
+    Retrieves or generates a submission ID based on the information in the provided workbook.
+
+    The function extracts the reporting round and programme ID from the workbook.
+    If a matching programme exists in the same reporting round, it calls the
+    delete_existing_submission function to delete the existing submission and its children.
+    The submission ID is then retrieved, and if there is no matching programme,
+    a new submission ID is generated.
+
+    :param workbook: a dictionary of DataFrames containing table data to determine the submission ID.
+    :return: a string representing the submission ID, and the Submission to delete
+    """
+    reporting_round = int(workbook["Submission_Ref"]["Reporting Round"].iloc[0])
+    programme_id = workbook["Programme_Ref"]["Programme ID"].iloc[0]
+    programme_exists_same_round = get_programme_by_id_and_round(programme_id, reporting_round)
+
+    submission_to_del = None
```

Comment:
can we move this to inside the else block, otherwise it gets re-assigned in the if block.